File: Home/home/Stater.cpp
#include "Stater.h"

Stater::Stater()
{
	_activeHigh = LOW;
	_btnState = !_activeHigh;  // initial button state in active-high logic
	_lastState = _btnState;
	_clickCount = 0;
	clicks = 0;
	depressed = 0;
	_lastBounceTime = 0;
	debounceTime = 20;            // Debounce timer in ms
	multiclickTime = 250;           // Time limit for multi clicks
	longClickTime = 1000;
	
	_editMode = LOW;
	_lastClickTime = 0;
	editModeOffTime = 3000;
}


void Stater::EditMode()
{	
	_editMode = HIGH;
	multiclickTime = 50;
	_lastClickTime = (long)millis();
}

void Stater::Update(boolean btnState)
{
	long now = (long)millis();      // get current time
	_btnState = btnState;  // current appearant button state

	// Make the button logic active-high in code
	if (!_activeHigh) _btnState = !_btnState;

	// If the switch changed, due to noise or a button press, reset the debounce timer
	if (_btnState != _lastState) _lastBounceTime = now;


	// debounce the button (Check if a stable, changed state has occured)
	if (now - _lastBounceTime > debounceTime && _btnState != depressed)
	{
		depressed = _btnState;
		if (depressed) _clickCount++;
	}

	// If the button released state is stable, report nr of clicks and start new cycle
	if (!depressed && (now - _lastBounceTime) > multiclickTime)
	{
		// positive count for released buttons
		clicks = _clickCount;
		_clickCount = 0;
	}

	// Check for "long click"
	if (depressed && (now - _lastBounceTime > longClickTime))
	{
		// negative count for long clicks
		clicks = 0 - _clickCount;
		_clickCount = 0;
	}
	
	//Off EditModel
	if(_editMode && (now - _lastClickTime > editModeOffTime)){
		_editMode = LOW;
		multiclickTime = 250;
	}

	_lastState = _btnState;
}

File: i2c-server/Stater.js
'use strict';


var debounceTime = 20;
var multiclickTime = 250;
var longClickTime = 1000;


var Stater = function (pin) { this.pin = pin; }
Stater.prototype._clickCallback = function () { };
Stater.prototype._doubleClickCallback = function () { };
Stater.prototype._longClickCallback = function () { };

Stater.prototype.clickCount = 0;
Stater.prototype.lastBounceTime = 0;
Stater.prototype.lastState = false;
Stater.prototype.depressed = false;

Stater.prototype.update = function (btnState) {
    var now = new Date().getTime();
    if (btnState != this.lastState) this.lastBounceTime = now;
    
    if (now - this.lastBounceTime > debounceTime && btnState != this.depressed) {
        this.depressed = btnState;
        if (this.depressed) {
            this.clickCount++;
        }
    }

    if (!this.depressed && (now - this.lastBounceTime) > multiclickTime) {
        if (this.clickCount === 1) {
            this._clickCallback(this.pin);
        } else if (this.clickCount === 2) {
            this._doubleClickCallback(this.pin);
        }
        this.clickCount = 0;
    }

    if (this.depressed && (now - this.lastBounceTime > longClickTime)) {
        this._longClickCallback(this.pin);
        this.clickCount = 0;
    }

    this.lastState = btnState;
}

Stater.prototype.onClick = function (callback) {
    this._clickCallback = callback;
}

Stater.prototype.onDoubleClick = function (callback) {
    this._doubleClickCallback = callback;
}

Stater.prototype.onLongClick = function (callback) {
    this._longClickCallback = callback;
}

module.exports = Stater;